### PR TITLE
Feature/sign elasticsearch http requests

### PR DESCRIPTION
**Pull request by NoelleDL**
### What

Added signing of requests to elasticsearch in the healthcheck and post request. https://trello.com/c/Sq3jAEtH/758-search-api-sign-elasticsearch-http-requests

### How to review

Check that code signs requests and all tests pass. 

### Who can review

Anyone but me.

### CODE REVIEW

File: cmd/dp-search-api/main.go [hunk 1]
```diff
@@ -50,7 +51,16 @@ func main() {
...
+	var esSigner *esauth.Signer
+	if cfg.SignElasticsearchRequests {
+		esSigner, err = esauth.NewAwsSigner("", "", cfg.AwsRegion, cfg.AwsService)
+		if err != nil {
+			log.Event(nil, "failed to create aws v4 signer", log.ERROR, log.Error(err))
+			os.Exit(1)
+		}
+	}
+
+	elasticSearchClient := elasticsearch.New(cfg.ElasticSearchAPIURL, dphttp.NewClient(), cfg.SignElasticsearchRequests, esSigner, cfg.AwsRegion, cfg.AwsService)
```
Comment (daiLlew) [inline cmd/dp-search-api/main.go:63]: Question:

Just checking if its OK for `esSigner` to be `nil` inside the `elasticseach.New` func? If it is nil is that handled or will it error? I assume its OK but best to double check.

File: elasticsearch/client.go [hunk 2]
```diff
@@ -71,7 +82,9 @@ func (cli *Client) post(ctx context.Context, index string, docType string, actio
 	}
 
 	if cli.signRequests {
-		awsauth.Sign(req)
+		if err = cli.awsSDKSigner.Sign(req, reader, time.Now()); err != nil {
+			return nil, err
+		}
```
Comment (daiLlew) [inline elasticsearch/client.go:87]: I am being picky as it shouldn't ever happen but I think  you should check that `awsSDKSigner` is not `nil` if `signRequests == true`. 

With the current code that shouldn't be possible _but_ if this was every refactored its possible that assumption could be missed and a `nil` value could slip through and cause pain. If we add the check we can give an explicit error message to state whats wrong.

File: elasticsearch/client_test.go [hunk 3]
```diff
@@ -144,7 +155,9 @@ func TestGetStatus(t *testing.T) {
 				},
 			}
 
-			client := New("http://localhost:999", dphttpMock, false)
+			var testSigner *esauth.Signer
+
+			client := New("http://localhost:999", dphttpMock, false, testSigner, "es", "eu-west-1")
```
Comment (daiLlew) [inline elasticsearch/client_test.go:160]: Can you also add a new test case to cover the `signRequests == true` && `signer = nil` please? 😄

File: cmd/dp-search-api/main.go (hunk 1 above)
Comment (NoelleDL) [reply cmd/dp-search-api/main.go:63]: Yes, see my message below.

File: elasticsearch/client.go (hunk 2 above)
Comment (NoelleDL) [reply elasticsearch/client.go:87]: This is being handled by the dp-elasticsearch library: https://github.com/ONSdigital/dp-elasticsearch/blob/main/awsauth/sign.go#L55. I've added a test to check that the correct error message is being returned.

**Review by daiLlew — APPROVED** [review]
